service/metric/middleware.go

```diff
@@ -0,0 +1,66 @@
+package metric
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// CollectAPIStats middleware collects different statistics from the HTTP request
+// use this middleware like this:
+// router := mux.NewRouter()
+// router.Handle("/v1/models", metric.CollectAPIStats("modelList", http.HandlerFunc(model.List))).Methods("GET")
+func CollectAPIStats(view string, inner http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ww := &recordResponse{ResponseWriter: w}
+		start := time.Now()
+
+		inner.ServeHTTP(ww, r)
+
+		// count all server side errors
+		if ww.status >= 500 {
+			HTTPIncomingErrorsCounterVec.WithLabelValues(r.Method, ww.Status(), view).Inc()
+		}
+		// count 504 Gateway Timeout
+		if ww.status == 504 {
+			HTTPIncomingTimeoutsCounterVec.WithLabelValues(r.Method, view).Inc()
+		}
+		// record prometheus metric for latency in ms
+		latency := float64(time.Since(start).Nanoseconds()) / 1000000
+
+		HTTPIncomingLatencyHistogramVec.WithLabelValues(r.Method, ww.Status(), view).Observe(latency)
+		HTTPIncomingRequestCounterVec.WithLabelValues(r.Method, ww.Status(), view).Inc()
+	})
+}
+
+// recordResponse is a proxy around an http.ResponseWriter
```

Comment:
This is the trick to get a response status from the already written response. `bytesWritten` are not used now, but we can add an additional metric for this information